grabbit/src/main/groovy/com/twcable/grabbit/client/batch/steps/workspace/DeleteBeforeWriteTasklet.groovy

```diff
@@ -0,0 +1,158 @@
+        }
+        else {
+            deletePartialTree(session)
+        }
+        session.save()
+        return RepeatStatus.FINISHED
+    }
+
+
+    private void deletePartialTree(final Session session) {
+        final relativeExcludePaths = excludePaths.collect { it - jobPath }
+        final relativeRoot = session.getNode(jobPath)
+        deletePartialTree(relativeRoot, relativeExcludePaths)
+    }
+
+
+    private void deletePartialTree(Node rootNode, Collection<String> relativeExcludePaths) {
+        //Base case.  If a root node does not have any exclude paths under it, it must have been excluded
+        if(!relativeExcludePaths) {
+            return
+        }
+        //Compute the current tree level excluded nodes, and compute the remaining paths to traverse
+        Collection<NodeAndExcludePaths> nodeAndExcludePaths = relativeExcludePaths.inject([] as Collection<NodeAndExcludePaths>) { def acc, def thisPath ->
+            final thisNodeAndExcludePath = NodeAndExcludePaths.fromPath(thisPath)
+            //If we have already created a nodeAndExcludePath for this node name, just add to it's exclude paths
+            final matchingNodeName = acc.find { it.nodeName == thisNodeAndExcludePath.nodeName }
+            if(matchingNodeName) {
+                matchingNodeName.excludePaths.addAll(thisNodeAndExcludePath.excludePaths)
+            }
+            else {
+                acc.add(thisNodeAndExcludePath)
+            }
+            return acc
+        } as Collection
+        //Delete nodes allowed under this tree
+        final rootNodeChildren = rootNode.getNodes()
+        while(rootNodeChildren.hasNext()) {
+            final currentNode = rootNodeChildren.nextNode()
+            //If this node is in our exclusion list, don't delete it
+            if(!(nodeAndExcludePaths.find { currentNode.name == it.nodeName })) {
+                currentNode.remove()
+            }
+        }
+        //Recurse on each
+        nodeAndExcludePaths.each {
+            deletePartialTree(rootNode.getNode(it.nodeName), it.excludePaths)
+        }
+    }
+
+    /**
+     * If we have a path for a node such as 'foo/bar/doo'
+     * node is 'foo', and the excludePaths is '/bar/doo'
+     */
+    static class NodeAndExcludePaths {
+
+        String nodeName
+        Collection<String> excludePaths
+
+        static NodeAndExcludePaths fromPath(final String path) {
+            //Remove leading '/'
+            final String thisPath = path.replaceFirst('/', '')
+            //We need to find the current node, and the remaining path e.g 'foo/bar/doo', node: foo, remaining path: /bar/doo
+            final slashIndex = thisPath.indexOf('/')
+            //There is no remaining path
+            if(slashIndex == -1) {
+                return new NodeAndExcludePaths(nodeName: thisPath, excludePaths: [])
+            }
+            final remainingPath = thisPath.substring(slashIndex)
+            final node = thisPath - remainingPath
```

Comment:
More efficient than the `minus(..)`
```groovy
final node = thisPath.substring(0, slashIndex)
```
also a bit clearer as to the intent and less likely to get caught on subtle bugs like `remainingPath` happening to be a substring in what is supposed to be `node`